hearthstone/hero_pool.py

```diff
@@ -244,15 +267,61 @@ def hero_power_impl(self, context: BuyPhaseContext):
 class QueenWagtoggle(Hero):
     power_cost = 1
 
-    def hero_power_impl(self, context: 'BuyPhaseContext'):
+    def hero_power_impl(self, context: 'BuyPhaseContext', board_index: Optional['BoardIndex'] = None,
+                        store_index: Optional['StoreIndex'] = None):
         for card in one_minion_per_type(context.owner.in_play, context.randomizer):
             card.attack += 2
 
 
 class ForestWardenOmu(Hero):
-    def hero_power_valid_impl(self, context: 'BuyPhaseContext'):
+    def hero_power_valid_impl(self, context: 'BuyPhaseContext', board_index: Optional['BoardIndex'] = None,
+                              store_index: Optional['StoreIndex'] = None):
         return False
 
     def handle_event(self, event: 'CardEvent', context: 'BuyPhaseContext'):
         if event.event is EVENTS.TAVERN_UPGRADE:
-            context.owner.coins = min(context.owner.coins + 2, 10)
+            context.owner.plus_coins(2)
+
+
+class GeorgeTheFallen(Hero):
+    power_cost = 3
+    power_target_location = CardLocation.BOARD
+
+    def hero_power_valid_impl(self, context: BuyPhaseContext, board_index: Optional['BoardIndex'] = None,
+                              store_index: Optional['StoreIndex'] = None):
+        return not context.owner.in_play[board_index].divine_shield
+
+    def hero_power_impl(self, context: 'BuyPhaseContext', board_index: Optional['BoardIndex'] = None,
+                        store_index: Optional['StoreIndex'] = None):
+        context.owner.in_play[board_index].divine_shield = True
+
+
+class RenoJackson(Hero):
+    power_cost = 0
+    power_target_location = CardLocation.BOARD
+
+    def hero_power_valid_impl(self, context: BuyPhaseContext, board_index: Optional['BoardIndex'] = None,
+                              store_index: Optional['StoreIndex'] = None):
+        return not context.owner.in_play[board_index].golden
+
+    def hero_power_impl(self, context: 'BuyPhaseContext', board_index: Optional['BoardIndex'] = None,
+                        store_index: Optional['StoreIndex'] = None):
+        context.owner.in_play[board_index].golden_transformation([])
+        self.can_use_power = False
+
+
+class JandiceBarov(Hero):
+    power_cost = 0
+    power_target_location = CardLocation.BOARD
+
+    def hero_power_valid_impl(self, context: BuyPhaseContext, board_index: Optional['BoardIndex'] = None,
+                              store_index: Optional['StoreIndex'] = None):
+        return not context.owner.in_play[board_index].golden and len(context.owner.store) > 0
+
+    def hero_power_impl(self, context: 'BuyPhaseContext', board_index: Optional['BoardIndex'] = None,
+                        store_index: Optional['StoreIndex'] = None):
+        board_minion = context.owner.in_play.pop(board_index)
+        store_minion = context.randomizer.select_from_store(context.owner.store)
+        context.owner.store.remove(store_minion)
+        context.owner.in_play.append(store_minion)
```

Comment:
Do we need to trigger some events here?  E.G.  the demon guy who gives invulnerability. Maybe we should have helper functions that move cards and trigger events, so we don't need to catch all of these individual cases.  Not sure, since there might a large number of different ways cards can move around.

We might also thing about organizing our events around cards moving from location X to location Y.

Reply:
Good idea. There is a ReturnToHand event, so I can change it to LeavePlay or something similar.

Alternatively, we could have immunity be some card attribute called give_immunity as opposed to a player attribute, and  have the player.take_damage() function check for cards with that attribute. That way we wouldn't even need a LeavePlay event.